src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs

```diff
@@ -2,218 +2,173 @@
-    /// <param name="handler">event handler to update for manager</param>
-    public void UpdateHandlerForManager(T manager, TU handler)
-    {
-        if (_concurrentManagerHandlerMap.ContainsKey(manager))
-        {
-            _concurrentManagerHandlerMap[manager] = handler;
-        }
-        else
+        _workloads.AddRange(workloads);
+
+        lock (_lock)
         {
-            AddManager(manager, handler);
+            ClearSlots();
+            RunWorkInParallel();
         }
     }
 
-    /// <summary>
-    /// Get the event handler associated with the manager.
-    /// </summary>
-    /// <param name="manager">Manager</param>
-    public TU GetHandlerForGivenManager(T manager)
+    private bool RunWorkInParallel()
     {
-        return _concurrentManagerHandlerMap[manager];
-    }
+        // TODO: Right now we don't re-use shared hosts, but if we did, this is the place
+        // where we should find a workload that fits the manager if any of them is shared.
+        // Or tear it down, and start a new one.
 
-    /// <summary>
-    /// Get total number of active concurrent manager
-    /// </summary>
-    public int GetConcurrentManagersCount()
-    {
-        return _concurrentManagerHandlerMap.Count;
-    }
+        if (_eventHandler == null)
+            throw new InvalidOperationException($"{nameof(_eventHandler)} was not provided.");
 
-    /// <summary>
-    /// Get instances of all active concurrent manager
-    /// </summary>
-    public IEnumerable<T> GetConcurrentManagerInstances()
-    {
-        return _concurrentManagerHandlerMap.Keys.ToList();
-    }
+        if (_getEventHandler == null)
+            throw new InvalidOperationException($"{nameof(_getEventHandler)} was not provided.");
 
+        if (_runWorkload == null)
+            throw new InvalidOperationException($"{nameof(_runWorkload)} was not provided.");
 
-    /// <summary>
-    /// Updates the Concurrent Executors according to new parallel setting
-    /// </summary>
-    /// <param name="newParallelLevel">Number of Parallel Executors allowed</param>
-    public void UpdateParallelLevel(int newParallelLevel)
-    {
-        if (_concurrentManagerHandlerMap == null)
+        List<SlotWorkloadPair> workToRun = new();
+        lock (_lock)
         {
-            // not initialized yet
-            // create rest of concurrent clients other than default one
-            _concurrentManagerHandlerMap = new ConcurrentDictionary<T, TU>();
-            for (int i = 0; i < newParallelLevel; i++)
+            if (_workloads.Count == 0)
+                return true;
+
+            var availableSlots = _managerSlots.Where(slot => slot.IsAvailable).ToList();
+            var availableWorkloads = _workloads.Where(workload => workload != null).ToList();
+            var amount = Math.Min(availableSlots.Count, availableWorkloads.Count);
+            var workloadsToRun = availableWorkloads.Take(amount).ToList();
```

Comment:
Looking back at the code and thinking why this does not break the functionality: It makes no difference if we take the workload from availableWorkloads, or from its "slice" workloadsToRun. They both start at index 0, and they both have the same items. workloadsToRun just has exactly the needed amount of items as `amount`, and availableWorkloads has also the same items (in the same order), and then more that we won't touch because they are on index higher than amount.

Reply:
I fixed the code to use workloadsToRun, it makes debugging slightly easier.